### PR TITLE
fix: private trie cache db with gc

### PR DESCRIPTION
**Pull request by baptiste-b-pegasys**
Don't save the private trie DB at every commit.

Use the same way as the public trie DB, where data stored are optimized with cache and GC.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -1719,27 +1734,44 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
 	} else {
 		// Full but not archive node, do proper garbage collection
 		triedb.Reference(root, common.Hash{}) // metadata reference to keep trie alive
 		bc.triegc.Push(root, -int64(block.NumberU64()))
 
+		// Quorum
+		if len(privateRoot.Bytes()) != 0 {
+			privateTrieDB.Reference(privateRoot, common.Hash{}) // metadata reference to keep private trie alive
+			bc.privateTrieGC.Push(privateRoot, -int64(block.NumberU64()))
+		}
```
Comment (ricardolyn) [inline core/blockchain.go:1761]: add and `// End Quorum`

File: core/blockchain.go [hunk 2]
```diff
@@ -1758,6 +1790,8 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 					}
 					// Flush an entire trie and restart the counters
 					triedb.Commit(header.Root, true, nil)
+					privateroot := rawdb.GetPrivateStateRoot(bc.db, header.Root)
```
Comment (ricardolyn) [inline core/blockchain.go:1801]: quorum comments around?

File: core/blockchain.go [hunk 3]
```diff
@@ -1719,27 +1734,44 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
-				nodes, imgs = triedb.Size()
-				limit       = common.StorageSize(bc.cacheConfig.TrieDirtyLimit) * 1024 * 1024
+				nodes, imgs               = triedb.Size()
+				privateNodes, privateImgs = privateTrieDB.Size()
+				limit                     = common.StorageSize(bc.cacheConfig.TrieDirtyLimit) * 1024 * 1024
 			)
 			if nodes > limit || imgs > 4*1024*1024 {
 				triedb.Cap(limit - ethdb.IdealBatchSize)
 			}
+			if privateNodes > limit || privateImgs > 4*1024*1024 {
```
Comment (ricardolyn) [inline core/blockchain.go:1776]: quorum comments around?

File: core/blockchain.go [hunk 4]
```diff
@@ -1771,6 +1805,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 				}
 				triedb.Dereference(root.(common.Hash))
 			}
+			for !bc.privateTrieGC.Empty() {
```
Comment (ricardolyn) [inline core/blockchain.go:1819]: quorum comments around?

File: core/blockchain.go [hunk 5]
```diff
@@ -1174,6 +1177,12 @@ func (bc *BlockChain) Stop() {
 				if err := triedb.Commit(recent.Root(), true, nil); err != nil {
 					log.Error("Failed to commit recent state trie", "err", err)
 				}
+				// Quorum
+				privateRoot := rawdb.GetPrivateStateRoot(bc.db, recent.Root())
+				log.Info("Writing private cached state to disk", "block", recent.Number(), "hash", recent.Hash(), "privateRoot", privateRoot)
+				if err := privateTrieDb.Commit(privateRoot, true, nil); err != nil {
+					log.Error("Failed to commit recent private state trie", "err", err)
+				}
```
Comment (ricardolyn) [inline core/blockchain.go:1185]: add `// End Quorum`

**Review by ricardolyn — DISMISSED** [review]
Looks good. only missing comments

File: core/blockchain.go [hunk 6]
```diff
@@ -1771,6 +1815,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 				}
 				triedb.Dereference(root.(common.Hash))
 			}
+			for !bc.privateTrieGC.Empty() {
```
Comment (baptiste-b-pegasys) [inline core/blockchain.go:1819]:
```suggestion
			// Quorum
			for !bc.privateTrieGC.Empty() {
```

File: core/blockchain.go [hunk 7]
```diff
@@ -1771,6 +1815,14 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
+				root, number := bc.privateTrieGC.Pop()
+				if uint64(-number) > chosen {
+					bc.privateTrieGC.Push(root, number)
+					break
+				}
+				privateTrieDB.Dereference(root.(common.Hash))
+			}
```
Comment (baptiste-b-pegasys) [inline core/blockchain.go:1826]:
```suggestion
			}
			// End Quorum
```

File: core/blockchain.go [hunk 8]
```diff
@@ -1719,27 +1735,49 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
+		}
+		// End Quorum
+
 		if current := block.NumberU64(); current > TriesInMemory {
 			// If we exceeded our memory allowance, flush matured singleton nodes to disk
 			var (
-				nodes, imgs = triedb.Size()
-				limit       = common.StorageSize(bc.cacheConfig.TrieDirtyLimit) * 1024 * 1024
+				nodes, imgs               = triedb.Size()
+				privateNodes, privateImgs = privateTrieDB.Size()
```
Comment (baptiste-b-pegasys) [inline core/blockchain.go]:
```suggestion
				privateNodes, privateImgs = privateTrieDB.Size() // Quorum
```

File: core/blockchain.go [hunk 9]
```diff
@@ -1185,9 +1195,15 @@ func (bc *BlockChain) Stop() {
 		for !bc.triegc.Empty() {
 			triedb.Dereference(bc.triegc.PopItem().(common.Hash))
 		}
+		for !bc.privateTrieGC.Empty() {
```
Comment (baptiste-b-pegasys) [inline core/blockchain.go]:
```suggestion
		for !bc.privateTrieGC.Empty() { // Quorum
```

File: core/blockchain.go [hunk 10]
```diff
@@ -1185,9 +1195,15 @@ func (bc *BlockChain) Stop() {
...
+			privateTrieDb.Dereference(bc.privateTrieGC.PopItem().(common.Hash))
+		}
 		if size, _ := triedb.Size(); size != 0 {
 			log.Error("Dangling trie nodes after full cleanup")
 		}
+		if size, _ := privateTrieDb.Size(); size != 0 {
```
Comment (baptiste-b-pegasys) [inline core/blockchain.go]:
```suggestion
		if size, _ := privateTrieDb.Size(); size != 0 { // Quorum
```

**Review by ricardolyn — APPROVED** [review]

**Review by nmvalera — APPROVED** [review]
LGTM

**Pull request merged**